sound/soc/sof/topology.c

```diff
@@ -3271,86 +3287,93 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 		return -EINVAL;
 	}
 
-	/* Send BE DAI link configurations to DSP */
-	memset(&config, 0, sizeof(config));
-
-	/* get any common DAI tokens */
-	ret = sof_parse_tokens(scomp, &config, dai_link_tokens,
-			       ARRAY_SIZE(dai_link_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret != 0) {
-		dev_err(scomp->dev, "error: parse link tokens failed %d\n",
-			le32_to_cpu(private->size));
-		return ret;
-	}
-
 	/*
 	 * DAI links are expected to have at least 1 hw_config.
 	 * But some older topologies might have no hw_config for HDA dai links.
 	 */
-	num_hw_configs = le32_to_cpu(cfg->num_hw_configs);
-	if (!num_hw_configs) {
-		if (config.type != SOF_DAI_INTEL_HDA) {
+	hw_config = cfg->hw_config;
+	num_conf = le32_to_cpu(cfg->num_hw_configs);
+	if (!num_conf) {
+		if (config->type != SOF_DAI_INTEL_HDA) {
 			dev_err(scomp->dev, "error: unexpected DAI config count %d!\n",
 				le32_to_cpu(cfg->num_hw_configs));
 			return -EINVAL;
 		}
+		num_conf = 1;
 	} else {
 		dev_dbg(scomp->dev, "tplg: %d hw_configs found, default id: %d!\n",
 			cfg->num_hw_configs, le32_to_cpu(cfg->default_hw_config_id));
 
-		for (i = 0; i < num_hw_configs; i++) {
-			if (cfg->hw_config[i].id == cfg->default_hw_config_id)
+		for (def_conf = 0; def_conf < num_conf; def_conf++) {
+			if (hw_config[def_conf].id == cfg->default_hw_config_id)
 				break;
 		}
 
-		if (i == num_hw_configs) {
+		if (def_conf == num_conf) {
 			dev_err(scomp->dev, "error: default hw_config id: %d not found!\n",
 				le32_to_cpu(cfg->default_hw_config_id));
 			return -EINVAL;
 		}
 	}
 
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* reserve memory for all hw configs, eventually freed by widget */
+	config = kzalloc(num_conf * sizeof(*config), GFP_KERNEL);
+	if (!config)
+		return -ENOMEM;
 
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.format = le32_to_cpu(hw_config->fmt);
+	/* Send BE DAI link configurations to DSP */
+	memset(config, 0, num_conf * sizeof(config));
+
+	/* get any common DAI tokens */
+	ret = sof_parse_tokens(scomp, config, dai_link_tokens,
+			       ARRAY_SIZE(dai_link_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		kfree(config);
+		dev_err(scomp->dev, "error: parse link tokens failed %d\n",
+			le32_to_cpu(private->size));
+		return ret;
```

Comment:
maybe use a `goto` and set up an error return path under `return 0;`

Reply:
I remember we we're trying to get rid of goto's? So I was on purpose checking that "ok it is 2 free calls in error cases, so I don't put goto"... But anyway I'm fine adding goto error path.

Reply:
There might be slightly different interpretations. I think what the kernel docs say is something along the lines of "avoid `goto`s in general, but ok to use where it improves the code, particularly for error handling." So, it's a general consensus that it's ok for error handling, otherwise it's a bit of a matter of personal judgement.